rcl/src/rcl/security.c

```diff
@@ -111,16 +111,16 @@ char * exact_match_lookup(
 {
   // Perform an exact match for the context name in directory <root dir>.
   char * secure_root = NULL;
+  char * contexts_dir = NULL;
+  contexts_dir = rcutils_join_path(ros_secure_root_env, "contexts", *allocator);
   // "/" case when root namespace is explicitly passed in
   if (0 == strcmp(name, "/")) {
-    secure_root = rcutils_strdup(ros_secure_root_env, *allocator);
+    secure_root = contexts_dir;
   } else {
     char * relative_path = NULL;
-    char * contexts_dir = NULL;
     // Get native path, ignore the leading forward slash
     // TODO(ros2team): remove the hard-coded length, use the length of the root namespace instead
     relative_path = rcutils_to_native_path(name + 1, *allocator);
-    contexts_dir = rcutils_join_path(ros_secure_root_env, "contexts", *allocator);
     secure_root = rcutils_join_path(contexts_dir, relative_path, *allocator);
     allocator->deallocate(relative_path, allocator->state);
     allocator->deallocate(contexts_dir, allocator->state);
```

Comment:
This should be moved to the same scope as the allocation, otherwise it's not deallocated for "/"

Reply:
why?

In the `if` case, the `contexts_dir` pointer is just being copied to `secure_root`, so deallocating it would deallocate the return value.
Only in the `else` case deallocation is needed.

Reply:
Yeah, in the "/" case, we actually do *not* want to deallocate it, since we want to return the heap-allocated pointer.  In all other cases, we want to use it and then deallocate it, because we'll be returning something else.  I don't love the asymmetry (it will be hard to follow later), but I do think it is correct.

Reply:
Oh you're totally right I read too fast